File: 02_Supervised_Learning-Regression/R/Code/05_polynomial_regression.R
#Polynomial regression

#Data Preprocessing ; importing the

dataset = read.csv('Position_Salaries.csv')
dataset = dataset[, 2:3]

#Splitting the dataset in to training set and test set
#install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$Salary, SplitRatio = 0.8 )
# training_set = subset(dataset, split == TRUE)
# test_set= subset(dataset, split == FALSE)

#Feature scaling
# training_set[, 2:3] = scale(training_set[, 2:3])
# test_set[, 2:3] = scale(test_set[, 2:3])

#Fitting the linear regression model to the dataset
lin_reg = lm(formula = Salary ~ .,
             data = dataset)

#Fitting the polynomial regression model to the dataset
dataset$Level2 = dataset$Level^2
dataset$Level3 = dataset$Level^3
#addding new polynomial degree to fit curve closer
dataset$Level4 = dataset$Level^4
poly_reg = lm(formula = Salary ~ .,
              data = dataset)


#Visualising the linear regression results
#install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            color = 'blue') +
  ggtitle("Truth or Bluff(Linear Regression)") +
  xlab("Level") +
  ylab("Salary")

#Visualising the polynomial regression results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            color = 'blue') +
  ggtitle("Truth or Bluff(Polynomail Regression)") +
  xlab("Level") +
  ylab("Salary")

#Predict a new result with linear regression 
y_pred = predict(lin_reg, newdata = data.frame(Level = 6.5))

#Predict a new result with polynomial regression
y_pred_poly = predict(poly_reg, newdata = data.frame(Level = 6.5,
                                                Level2 = 6.5^2,
                                                Level3 = 6.5^3,
                                                Level4 = 6.5^4))

File: 02_Supervised_Learning-Regression/Python/Code/05_polynomial_regression.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu Nov 24 19:30:26 2016

@author: pavan.govindraj
"""

# Data preprocessing => import the dataset
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

dataset = pd.read_csv('Position_Salaries.csv')

X = dataset.iloc[:, 1:2].values
Y = dataset.iloc[:, 2].values


#Splitting the dataset in to training set and test set
#No training set and test set splitting as max info is required 
'''from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,Y,test_size = 0.2, 
                                                    random_state = 0)
'''

#Feature scaling
"""
#May be required for some dataset 
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""

#Fitting the linear regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X,Y)

#Fitting the polynomial regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4) # we can use more degree for precise
X_poly = poly_reg.fit_transform(X)
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, Y)

#Visualising the linear regression
plt.scatter(X, Y, color = 'red' )
plt.plot(X,lin_reg.predict(X), color = 'blue')
plt.title("Truth or Bluff(Linear Regression)")
plt.xlabel("Position")
plt.ylabel("Salary")
plt.show()

#Visualising the polynomial regression
plt.scatter(X, Y, color = 'red' )
plt.plot(X,lin_reg_2.predict(poly_reg.fit_transform(X)), color = 'blue')
plt.title("Truth or Bluff(Polynomial Regression)")
plt.xlabel("Position")
plt.ylabel("Salary")
plt.show()

#Improving the polynomial prediction, making the curve smoooth than lines
plt.scatter(X, Y, color = 'red' )
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.plot(X_grid,lin_reg_2.predict(poly_reg.fit_transform(X_grid)), color = 'blue')
plt.title("Truth or Bluff(Polynomial Regression)")
plt.xlabel("Position")
plt.ylabel("Salary")
plt.show()

#Predicting the new value using linear regression
lin_reg.predict(6.5)

#Predicting the new value using polynomial regression
lin_reg_2.predict(poly_reg.fit_transform(6.5))

